### PR TITLE
Eval function for string_to_lower/upper_case builtin functions

### PR DESCRIPTION
**Pull request by romainbrenguier**

### CODE REVIEW

File: src/solvers/refinement/string_builtin_function.cpp [hunk 1]
```diff
@@ -187,6 +187,25 @@ exprt string_set_char_builtin_functiont::length_constraint() const
...
+{
+  auto input_opt = eval_string(input, get_value);
+  if(!input_opt)
+    return {};
+  for(mp_integer &c : input_opt.value())
+  {
+    if(
+      ('A' <= c && c <= 'Z') || (0xc0 <= c && c <= 0xd6) ||
+      (0xd8 <= c && c <= 0xde))
+      c += 0x20;
```
Comment (peterschrammel) [inline src/solvers/refinement/string_builtin_function.cpp]: braces preferred

File: src/solvers/refinement/string_builtin_function.h [hunk 2]
```diff
@@ -177,6 +177,37 @@ class string_set_char_builtin_functiont
   exprt length_constraint() const override;
 };
 
+class string_to_lower_case_builtin_functiont
```
Comment (peterschrammel) [inline src/solvers/refinement/string_builtin_function.h]: doxygen

File: src/solvers/refinement/string_builtin_function.cpp [hunk 3]
```diff
@@ -206,6 +206,23 @@ optionalt<exprt> string_to_lower_case_builtin_functiont::eval(
...
 
+optionalt<exprt> string_to_upper_case_builtin_functiont::eval(
+  const std::function<exprt(const exprt &)> &get_value) const
+{
+  auto input_opt = eval_string(input, get_value);
+  if(!input_opt)
+    return {};
+  for(mp_integer &c : input_opt.value())
+  {
+    if('a' <= c && c <= 'z')
```
Comment (peterschrammel) [inline src/solvers/refinement/string_builtin_function.cpp]: why don't you consider the same set of characters as in lower_case?

File: src/solvers/refinement/string_builtin_function.h [hunk 4]
```diff
@@ -208,6 +208,37 @@ class string_to_lower_case_builtin_functiont
   };
 };
 
+class string_to_upper_case_builtin_functiont
```
Comment (peterschrammel) [inline src/solvers/refinement/string_builtin_function.h]: doxygen

File: src/solvers/refinement/string_constraint_generator_main.cpp [hunk 5]
```diff
@@ -456,10 +456,6 @@ exprt string_constraint_generatort::add_axioms_for_function_application(
     res=add_axioms_for_compare_to(expr);
   else if(id==ID_cprover_string_literal_func)
     res=add_axioms_from_literal(expr);
-  else if(id==ID_cprover_string_concat_func)
```
Comment (peterschrammel) [inline src/solvers/refinement/string_constraint_generator_main.cpp:459]: typo in commit message: redundant

File: jbmc/regression/jbmc-strings/StringToLowerCase/Test.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
+public class Test {
+    public String det()
+    {
+        StringBuilder builder = new StringBuilder();
+        builder.append("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ".toLowerCase());
+        builder.append("abcdeghijklmnopfrstuqwxyzABCDvFGHIJKLMENOPQRSTUVWXYZ".toLowerCase());
```
Comment (peterschrammel) [inline jbmc/regression/jbmc-strings/StringToLowerCase/Test.java:6]: do you have tests for the other letters too?

File: src/solvers/refinement/string_builtin_function.cpp (hunk 3 above)
Comment (romainbrenguier) [reply src/solvers/refinement/string_builtin_function.cpp]: it's to mimic the function encoded by the constraints for these two functions. Maybe I should adapt the constraints for the to_upper_case function to work for the same characters as lower_case, otherwise that can be confusing.

File: jbmc/regression/jbmc-strings/StringToLowerCase/Test.java (hunk 6 above)
Comment (romainbrenguier) [reply jbmc/regression/jbmc-strings/StringToLowerCase/Test.java:6]: No, but we don't test the result here anyway, we are just testing that unnecessary constraints are not added. I didn't find a good way to test it here: we cannot add assertions because that could change the result and recovering the string from the trace is difficult.

File: src/solvers/refinement/string_constraint_generator_main.cpp (hunk 5 above)
Comment (romainbrenguier) [reply src/solvers/refinement/string_constraint_generator_main.cpp:459]: fixed

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 15c01deec9bf0957a7d316a2ba29afd7db2f43c3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/81007322
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

**Review by peterschrammel — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 182256ac8f8e0bfb64fbd7a01db62bf80f486529).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/81011045

File: src/solvers/refinement/string_constraint_generator.h [hunk 7]
```diff
@@ -175,6 +175,9 @@ class string_constraint_generatort final
     const array_string_exprt &str,
     const exprt &position,
     const exprt &character);
+  exprt add_axioms_for_to_lower_case(
+    const array_string_exprt &res,
+    const array_string_exprt &str);
```
Comment (allredj) [inline src/solvers/refinement/string_constraint_generator.h:180]: Why exported?

File: src/solvers/refinement/string_builtin_function.cpp (hunk 1 above)
Comment (allredj) [reply src/solvers/refinement/string_builtin_function.cpp]: These codes don't mean anything to me. Could you perhaps add some comments, or better, define meaningful constant names for them?

File: src/solvers/refinement/string_builtin_function.cpp [hunk 8]
```diff
@@ -187,6 +187,25 @@ exprt string_set_char_builtin_functiont::length_constraint() const
     equal_exprt(return_code, return_value));
 }
 
+optionalt<exprt> string_to_lower_case_builtin_functiont::eval(
+  const std::function<exprt(const exprt &)> &get_value) const
+{
+  auto input_opt = eval_string(input, get_value);
```
Comment (allredj) [inline src/solvers/refinement/string_builtin_function.cpp]: Not related to this PR, but `eval_string` could benefit from some documentation.

File: src/solvers/refinement/string_builtin_function.cpp [hunk 9]
```diff
@@ -187,6 +187,25 @@ exprt string_set_char_builtin_functiont::length_constraint() const
     equal_exprt(return_code, return_value));
 }
 
+optionalt<exprt> string_to_lower_case_builtin_functiont::eval(
```
Comment (allredj) [inline src/solvers/refinement/string_builtin_function.cpp]: I don't understand what the role of the `get_value` function is. Could you please document? (probably in `string_builtin_functiont`.

File: src/solvers/refinement/string_builtin_function.cpp (hunk 1 above)
Comment (allredj) [reply src/solvers/refinement/string_builtin_function.cpp]: OK, I see you've added some documentation. it seems I shouldn't have reviewed commit by commit. I still think defining constants would be cleaner.

File: src/solvers/refinement/string_constraint_generator_transformation.cpp [hunk 10]
```diff
@@ -323,19 +331,14 @@ exprt string_constraint_generatort::add_axioms_for_to_lower_case(
...
 /// These axioms are:
-///   1. \f$ |{\tt res}| = |{\tt str}| \f$
-///   2. \f$ \forall i<|{\tt str}|.\ 'a'\le {\tt str}[i]\le 'z'
-///          \Rightarrow {\tt res}[i]={\tt str}[i]+'A'-'a' \f$
-///   3. \f$ \forall i<|{\tt str}|.\ \lnot ('a'\le {\tt str}[i] \le 'z')
-///          \Rightarrow {\tt res}[i]={\tt str}[i] \f$
-/// Note that index expressions are only allowed in the body of universal
-/// axioms, so we use a trivial premise and push our premise into the body.
+///   1. res.length = str.length
+///   2. forall i< str.length.
```
Comment (allredj) [inline src/solvers/refinement/string_constraint_generator_transformation.cpp]: nit: missing space before `<`

File: src/solvers/refinement/string_constraint_generator.h [hunk 11]
```diff
@@ -178,6 +178,9 @@ class string_constraint_generatort final
   exprt add_axioms_for_to_lower_case(
     const array_string_exprt &res,
     const array_string_exprt &str);
+  exprt add_axioms_for_to_upper_case(
+    const array_string_exprt &res,
+    const array_string_exprt &expr);
```
Comment (allredj) [inline src/solvers/refinement/string_constraint_generator.h]: why exported?

File: src/solvers/refinement/string_builtin_function.h [hunk 12]
```diff
@@ -210,6 +210,8 @@ class string_to_lower_case_builtin_functiont
   };
 };
 
+/// Converting each lowercase character of Basic Latin and Latin-1 supplement
```
Comment (allredj) [inline src/solvers/refinement/string_builtin_function.h]: Nit: `Converting` -> `converting`.

File: jbmc/regression/jbmc-strings/StringToUpperCase/test_dependency.desc [hunk 13]
```diff
@@ -0,0 +1,10 @@
+CORE
+Test.class
+--function Test.withDependency --verbosity 10 --max-nondet-string-length 10000
+^EXIT=10$
+^SIGNAL=0$
+assertion at file Test.java line 44 .*: SUCCESS
+assertion at file Test.java line 46 .*: FAILURE
+--
+--
+Check that when there are dependency, axioms are added correctly.
```
Comment (allredj) [inline jbmc/regression/jbmc-strings/StringToUpperCase/test_dependency.desc]: Ni: `dependency` -> `dependencies`

File: jbmc/regression/jbmc-strings/StringToUpperCase/Test.java [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+        builder.append("abcdfghijklmnopqrstuvwxyzABCDEFGHIJeKLMNOPQRSTUVWXYZ".toUpperCase());
+        return builder.toString();
+    }
+
+    public String nonDet(String s)
+    {
+        if(s.length() < 20)
+            return "Short string";
+        if(!s.startsWith("A"))
+            return "String not starting with A";
```
Comment (allredj) [inline jbmc/regression/jbmc-strings/StringToUpperCase/Test.java]: I guess you planned to exclude lower case `a` here.

**Review by allredj — APPROVED** [review]
Looks fine to me, if a bit painful to review with all these commits that sometimes overwrite the same lines multiple times. Fixup commits can be useful for the first reviewer to check that all change request  have been implemented, but are a nuisance for the second reviewer.

File: jbmc/regression/jbmc-strings/StringToLowerCase/Test.java (hunk 6 above)
Comment (martin-cs) [reply jbmc/regression/jbmc-strings/StringToLowerCase/Test.java:6]: A general, unhelpful point -- are we doing ASCII case conversion or Unicode case conversion, because the later, frankly, fills me with fear:

https://unicode.org/faq/casemap_charprop.html
https://www.unicode.org/reports/tr21/tr21-5.html

File: src/solvers/refinement/string_builtin_function.cpp [hunk 15]
```diff
@@ -187,6 +187,46 @@ exprt string_set_char_builtin_functiont::length_constraint() const
     equal_exprt(return_code, return_value));
 }
 
+static bool eval_is_upper_case(const mp_integer &c)
+{
+  return ('A' <= c && c <= 'Z') || (0xc0 <= c && c <= 0xd6) ||
+         (0xd8 <= c && c <= 0xde);
```
Comment (martin-cs) [inline src/solvers/refinement/string_builtin_function.cpp]: This says to me we're not doing ASCII.  Might you add a comment to say where these numbers come from?  Also, is there a reason to not use isupper et al.?

File: src/solvers/refinement/string_builtin_function.cpp [hunk 16]
```diff
@@ -187,6 +187,46 @@ exprt string_set_char_builtin_functiont::length_constraint() const
...
+optionalt<exprt> string_to_lower_case_builtin_functiont::eval(
+  const std::function<exprt(const exprt &)> &get_value) const
+{
+  auto input_opt = eval_string(input, get_value);
+  if(!input_opt)
+    return {};
+  for(mp_integer &c : input_opt.value())
+  {
+    if(eval_is_upper_case(c))
+      c += 0x20;
```
Comment (martin-cs) [inline src/solvers/refinement/string_builtin_function.cpp:212]: Does that always work?  Does it work better than to_lower?

File: src/solvers/refinement/string_builtin_function.cpp [hunk 17]
```diff
@@ -187,6 +187,46 @@ exprt string_set_char_builtin_functiont::length_constraint() const
...
+
+optionalt<exprt> string_to_upper_case_builtin_functiont::eval(
+  const std::function<exprt(const exprt &)> &get_value) const
+{
+  auto input_opt = eval_string(input, get_value);
+  if(!input_opt)
+    return {};
+  for(mp_integer &c : input_opt.value())
+  {
+    if(eval_is_upper_case(c - 0x20))
```
Comment (martin-cs) [inline src/solvers/refinement/string_builtin_function.cpp:228]: Same again.

File: src/solvers/refinement/string_builtin_function.cpp [hunk 18]
```diff
@@ -187,6 +187,46 @@ exprt string_set_char_builtin_functiont::length_constraint() const
     equal_exprt(return_code, return_value));
 }
 
+static bool eval_is_upper_case(const mp_integer &c)
+{
+  return ('A' <= c && c <= 'Z') || (0xc0 <= c && c <= 0xd6) ||
+         (0xd8 <= c && c <= 0xde);
+}
+
+optionalt<exprt> string_to_lower_case_builtin_functiont::eval(
```
Comment (martin-cs) [inline src/solvers/refinement/string_builtin_function.cpp:203]: I'm probably not paying attention, but why not just return the exprt anyway?  This is how the rewriter works.

**Review by martin-cs — COMMENTED** [review]
I'm unclear on what the "right" answers should be.  Is this for ASCII (in which case some parts are not needed), Unicode (in which case I am not qualified to comment but it doesn't look compliant), some Java standard (likewise), etc. ?

File: src/solvers/refinement/string_constraint_generator.h (hunk 7 above)
Comment (romainbrenguier) [reply src/solvers/refinement/string_constraint_generator.h:180]: We are going to use it in the string_builtin_function object evaluation function

File: src/solvers/refinement/string_builtin_function.cpp (hunk 15 above)
Comment (romainbrenguier) [reply src/solvers/refinement/string_builtin_function.cpp]: Yes I documented that so that it should be clearer now. We do a bit more that ASCII but not all unicode (that dould be quite complicated indeed). My idea is that we will provide this builtin function accessible as `CProverString.toUpperCase`, which only coincide with String.toUpperCase for a fragment of unicode (basic latin and latin-1 supplement for now), you can then provide a model of String.toUpperCase which either over-approximate (useful for verification) or under-approximate (useful for test generation) the behaviour of the JDK function.

File: src/solvers/refinement/string_builtin_function.cpp (hunk 16 above)
Comment (romainbrenguier) [reply src/solvers/refinement/string_builtin_function.cpp:212]: It mimics the behavior of the function specified by the constraints, which is not exactly the same as to_lower

File: src/solvers/refinement/string_builtin_function.cpp (hunk 18 above)
Comment (romainbrenguier) [reply src/solvers/refinement/string_builtin_function.cpp:203]: If the eval function is not successful, the solver will get the valuation in another way (calling get of the underlying solver and doing some post-processing).

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: a44468bea3d1ff3a482bb3528c0cb0058253f549).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/81049758

File: src/solvers/refinement/string_builtin_function.cpp (hunk 15 above)
Comment (tautschnig) [reply src/solvers/refinement/string_builtin_function.cpp]: This sounds scary. Please add tests and documentation (really, both!) that very precisely document the boundaries of what is/is not covered. I accept that there may not be enough time to do a perfect job (cover all of unicode), but then there needs to be clear documentation of what has (not) been done to avoid surprises.

File: src/solvers/refinement/string_builtin_function.cpp (hunk 18 above)
Comment (martin-cs) [reply src/solvers/refinement/string_builtin_function.cpp:203]: OK... is this called before or after solving?

File: src/solvers/refinement/string_builtin_function.cpp (hunk 15 above)
Comment (martin-cs) [reply src/solvers/refinement/string_builtin_function.cpp]: Yes; please, what @tautschnig said.

File: src/solvers/refinement/string_builtin_function.cpp (hunk 18 above)
Comment (romainbrenguier) [reply src/solvers/refinement/string_builtin_function.cpp:203]: It's called from `get` at the end of each string refinement step after solving by the underlying solver.

File: jbmc/regression/jbmc-strings/StringToLowerCase/Test.java (hunk 6 above)
Comment (owen-mc-diffblue) [reply jbmc/regression/jbmc-strings/StringToLowerCase/Test.java:6]: I seem to recall seeing tests for upper-casing non-ascii characters...
